app/src/androidTest/java/com/github/polypoly/app/network/RemoteDBTest.kt

```diff
@@ -204,4 +209,268 @@ class RemoteDBTest: PolyPolyTest(false, false) {
         val dataFound = remoteDB.getValue<User>(key).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
         assertEquals(data, dataFound)
     }
+
+    // =============================================================== REMOVE VALUE
+    @Test
+    fun unregisteredDataCantBeRemoved() {
+        val key = "some_key"
+        assertFalse(
+            remoteDB.removeValue(key).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        )
+    }
+
+    @Test
+    fun registeredDataCanBeRemoved() {
+        val key = "some_key"
+        val data = TEST_USER_1
+        remoteDB.registerValue(key, data).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        assertTrue(
+            remoteDB.removeValue(key).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        )
+    }
+
+    @Test
+    fun registeredDataIsRemoved() {
+        val key = "some_key"
+        val data = TEST_USER_1
+        remoteDB.registerValue(key, data).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        remoteDB.removeValue(key).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        assertFalse(
+            remoteDB.keyExists(key).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        )
+    }
+
+    // =============================================================================
+    // ========================================================= LISTENERS
+    // =============================================================================
+
+    // ========================================================= ADD ON CHANGE LISTENER
+    @Test
+    fun addingOnChangeListenerToUnregisteredDataFails() {
+        val key = "some_key"
+        val action = { _: User -> }
+        assertFalse(
+            remoteDB.addOnChangeListener(key, "tag", action, User::class).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        )
+    }
+
+    @Test
+    fun addingOnChangeListenerToRegisteredDataWorks() {
+        val key = "some_key"
+        val action = { _: User -> }
+
+        val data = TEST_USER_1
+        addDataToDB(data, key)
+
+        assertTrue(
+            remoteDB.addOnChangeListener(key, "tag", action, User::class).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        )
+    }
+
+    @Test
+    fun onChangeListenerIsExecutedAtUpdate() {
+        val key = "some_key"
+        var num = 0
+        val action = { _: User -> num += 1 }
+
+        val data1 = TEST_USER_1
+        val data2 = TEST_USER_2
+        addDataToDB(data1, key)
+
+        remoteDB.addOnChangeListener(key, "tag", action, User::class).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        remoteDB.updateValue(key, data2).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        remoteDB.updateValue(key, data1).get(TIMEOUT_DURATION, TimeUnit.SECONDS)
+        assertTrue(num >= 2)
```

Comment:
check that the callback is executed with the right data? you could replace the counter by a list and use Hamcrest's `contains` matcher

Reply:
I now added a test, thanks :)